compiler/rustc_target/src/asm/mod.rs

```diff
@@ -473,6 +471,17 @@ impl fmt::Display for InlineAsmRegOrRegClass {
     }
 }
 
+/// Type constrains for a particular register class.
+#[derive(Copy, Clone, Debug, Eq, PartialEq)]
+pub enum InlineAsmSupportedTypes {
+    /// Any type is allowed.
```

Comment:
Not exactly: this is still limited to primitive types and SIMD vectors.

By the way, is SPIR-V able to handle more exotic types such as `i128`?

Reply:
> By the way, is SPIR-V able to handle more exotic types such as `i128`?

For a general background: theoretically yes, integers are declared using an arbitrary bit width (like llvm), and the spec doesn't disallow (or even mention) 128 bit integers. Practically no, though, considering that there are extensions to enable support for 8 bit, 16 bit, and 64 bit integers, which implies no support for 128 bit (the SPIR-V world is full of weird unsaid implications like this).

However, I think in this case, 128 bit integers will either be disallowed outside `asm!` and so cannot leak into `asm!` and cause problems, or, they're represented as a pair of `u64` or whatever, and so still won't cause `asm!` problems with unrepresentable types.

> Not exactly: this is still limited to primitive types and SIMD vectors.

Yeah, we'd like to allow any type here, not sure if that's easiest to do in this PR or a follow-up (or if it's even possible to allow any type). It's certainly a little weird having an IR in `asm!`, arbitrary types are just... regular SSA IDs that look exactly like primitive types.

Reply:
+1 to allowing aggregate `Sized` types - I've run into this explicitly whilst working on https://github.com/EmbarkStudios/rust-gpu/issues/8 and it's blocking for some of the stuff I'd like to do there. Things get a lot trickier if we need to somehow deconstruct the type into its constituents, and this would also be a pessimisation for the generated code.